### PR TITLE
austin, austin-python, austin-tui: init

### PR DESCRIPTION
**Pull request by delroth**
###### Description of changes

https://github.com/NixOS/nixpkgs/issues/198752

Austin is a stack trace sampling profiler for CPython applications.

Submitted as draft for now since it requires #198802 (first commit of this PR).

out

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by aarmn** [comment]
Still waiting for the merge, should we make it final? once I tried to merge as well but usually drafts are ignored (as far as I know), If needed, you can convert it back to draft

**Comment by delroth** [comment]
This is staying in draft at least until the pythonRelaxDepsHook change
makes it to master (i.e. next staging-next merge), then I have a few things
to debug in terms of unit tests not passing on ofborg.

**Comment by aarmn** [comment]
Well now that `pythonRelaxDepsHook` is merged, can this become finalized?

File: pkgs/development/python-modules/austin-python/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/austin-python/default.nix:5]:
```suggestion
```

File: pkgs/development/python-modules/austin-python/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/austin-python/default.nix:40]:
```suggestion
```
We can't do this in pythonPackages.

File: pkgs/development/python-modules/austin-python/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/austin-python/default.nix:62]:
```suggestion
  doCheck = false;
```

we must use something like https://github.com/NixOS/nixpkgs/blob/c4c1866143ec02016d30eabed22cd6d0de39a10b/pkgs/development/python-modules/automat/default.nix#L50

File: pkgs/development/python-modules/austin-python/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/austin-python/default.nix:70]: Please short this comment

File: pkgs/development/python-modules/austin-tui/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/austin-tui/default.nix:3]:
```suggestion
```

File: pkgs/development/tools/profiling/austin/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/profiling/austin/default.nix:41]:
```suggestion
    austin-python
```

File: pkgs/development/tools/profiling/austin/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/profiling/austin/default.nix:44]:
```suggestion
    pytestCheckHook
```

File: pkgs/development/tools/profiling/austin/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/profiling/austin/default.nix:51]: Please adjust this to pytestCheckHook

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/austin-python/default.nix
Comment (delroth) [reply pkgs/development/python-modules/austin-python/default.nix:40]: I'm very confused: why? Clearly we can, since this PR as a whole is working fine, so you must be talking about a policy I don't know about. What is that policy?

Patching packages to find executable dependencies from the Nix store instead of relying on PATH lookup is completely standard for Nix AFAIK. Otherwise the only way to use this library will be to have `  austin` in $PATH, since the library is just a wrapper around executing the C program.

File: pkgs/development/tools/profiling/austin/default.nix
Comment (delroth) [reply pkgs/development/tools/profiling/austin/default.nix:51]: How do you provide the requisite environment variables to pytest then? I've read the hook's source code and it doesn't seem to support this. I could `export` in preCheckPhase but then the environment leaks unless I also restore in postCheckPhase, and that seems like a lot of effort for very marginal wins.

File: pkgs/development/python-modules/austin-python/default.nix
Comment (delroth) [reply pkgs/development/python-modules/austin-python/default.nix:5]: (Let's dedup with the thread about $PATH lookup since this only exists as a direct consequence to it.)

File: pkgs/development/python-modules/austin-python/default.nix
Comment (delroth) [reply pkgs/development/python-modules/austin-python/default.nix:62]: (Let's dedup with the thread about $PATH lookup since this only exists as a direct consequence to it.)

File: pkgs/development/tools/profiling/austin/default.nix
Comment (delroth) [reply pkgs/development/tools/profiling/austin/default.nix:41]: (Let's dedup with the thread about $PATH lookup since this only exists as a direct consequence to it.)

File: pkgs/development/python-modules/austin-python/default.nix
Comment (delroth) [reply pkgs/development/python-modules/austin-python/default.nix:70]: This might be the first time I'm being told in a code review to have *less* documentation for my code... What's your threshold? 3 lines is ok but 4 isn't?

File: pkgs/development/python-modules/austin-python/default.nix
Comment (samueldr) [reply pkgs/development/python-modules/austin-python/default.nix:40]: @SuperSandro2000 what is it that can't be done?

This here makes the package free-standing, self-sufficient and hermetic PATH-wise.

A stray `austin` binary in the PATH won't override the one the package is expected to be coherent with.

In fact, [looking at the implementation](https://github.com/P403n1x87/austin-python/blob/107db3b7fd88dafc35172f1b3eb4b15b7ed74ae0/austin/__init__.py#L286), I'd go as far as arguing the whole function should be stubbed out with the hardcoded path, but I don't know enough about the use case of providing multiple methods to find the binary here. (Which seems odd anyway...)

Comment (delroth) [reply pkgs/development/python-modules/austin-python/default.nix:40]: @samueldr I considered it but just modified the fallback instead -- the deliberate ways of pointing to another binary are hard to accidentally trigger, and unit testing relies on it too. It's a smaller diff too so all things considered... I think there are good arguments in favor of the current implementation.

Comment (SuperSandro2000) [reply pkgs/development/python-modules/austin-python/default.nix:40]: > PR as a whole is working fine

It is right now but pip is very picky about its inputs and if two of them have the same package name but other store paths you will receive an error. If this wouldn't be in place we could end up with the package being in the build inputs which would create an empty output.

This means we should reduce the amount of options for pythonPackages to a bare minimum.

File: pkgs/development/python-modules/austin-python/default.nix
Comment (SuperSandro2000) [reply pkgs/development/python-modules/austin-python/default.nix:70]:
```suggestion
    # the path to the austin binary is hardcoded, so it cannot fail to find it in PATH and the tests always fail
```

The problem is not about the length but the information density.

File: pkgs/development/tools/profiling/austin/default.nix
Comment (SuperSandro2000) [reply pkgs/development/tools/profiling/austin/default.nix:51]: > I could `export` in preCheckPhase

```
preCheck = ''
  export NIX_HARDENING_ENABLE=""
  export NIX_CFLAGS_COMPILE="$NIX_CFLAGS_COMPILE -fPIC"
'';
```

> then the environment leaks to everything running later in the derivation

That is not a problem. I don't think there is running anything after checkPhase. So tests do not modify $out.

File: pkgs/development/python-modules/austin-python/default.nix
Comment (SuperSandro2000) [reply pkgs/development/python-modules/austin-python/default.nix:62]: We can escape the infinity recursion here by disabling checks instead of overwriting a package to null. This also removes the antipattern and things like austin.version do not break.

Comment (delroth) [reply pkgs/development/python-modules/austin-python/default.nix:62]: I don't see how this is supposed to work:

- austin requires austin-python for its checkPhase
- austin-python needs to be patched to point to austin for it to be freestanding in nixpkgs

If you just make austin depend on austin-python minus its own tests, there's still a circular dependency because of the reference in postPatch.

File: pkgs/development/python-modules/austin-python/default.nix
Comment (delroth) [reply pkgs/development/python-modules/austin-python/default.nix:40]: I understand the theoretical problem you're referring to, but it very obviously does not apply to this situation. There is no reason to override this parameter other than to break the circular dependency, and it's documented as such. The likelihood of two conflicting versions ending up in the same closure is pretty much 0.

The exact same thing could be done without adding options too, e.g.  `austin` depending on `austin-python.override { austin = some-dummy-derivation; }` (instead of null), it's just uglier and hackier and doesn't properly document the circular dependency problem as part of the `austin-python` derivation.

File: pkgs/development/python-modules/austin-python/default.nix
Comment (delroth) [reply pkgs/development/python-modules/austin-python/default.nix:70]: I find this less readable, and this loses nuance that I intended to document:

- The path to the austin binary is hardcoded *by our patching*
- The tests always fail *but it's expected and understood to not be a problem*

File: pkgs/development/tools/profiling/austin/default.nix
Comment (delroth) [reply pkgs/development/tools/profiling/austin/default.nix:51]: How do you know that nothing runs in postCheck? How do you know nothing will run in postCheck in the future? This is more fragile for roughly 0 benefits because in this instance it's trivial to run pytest manually.

Happy to migrate this to pytestCheckHook once it supports the features that it would need to support to make using it here not fragile. Right now it would be leaking environment to an unknown amount of other hooks. FWIW installPhase and fixupPhase both run after checkPhase.

File: pkgs/development/python-modules/austin-python/default.nix
Comment (SuperSandro2000) [reply pkgs/development/python-modules/austin-python/default.nix:62]: We can break this by only running tests for austin in its passthru.

File: pkgs/development/tools/profiling/austin/default.nix
Comment (SuperSandro2000) [reply pkgs/development/tools/profiling/austin/default.nix:51]: > How do you know that nothing runs in postCheck?

I look at the source code of the hook and that it does not add anything there https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/hooks/pytest-check-hook.sh#L59-L74

> How do you know nothing will run in postCheck in the future?

You don't but if we ever add something there it will be carefully be considered. It is unlikely because it wouldn't run when checkPhase is overwritten. I could imagine adding some cleanup for pytestCheckHook there in the future if we ever need it. Nothing you need to worry about.

> This is more fragile for roughly 0 benefits because in this instance it's trivial to run pytest manually.

It is not fragile at all. A lot of packages use it and it is proven to make python maintenance easier by not modifying pytest invocations in packages.

> Right now it would be leaking environment to an unknown amount of other hooks.

We are not compiling anything after checkPhase so that is not a problem.

> FWIW installPhase and fixupPhase both run after checkPhase.

that is true for mkDerivation but [mkPythonPackages maps checkPhase to installCheckPhase](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/mk-python-derivation.nix#L197) so [only distPhase is run afterwards](https://github.com/NixOS/nixpkgs/blob/192e7bbb9240d8ceac534dd3b6ead7aa175dc3ce/doc/stdenv/stdenv.chapter.md#variables-affecting-phase-control-variables-affecting-phase-control)

**Comment by delroth** [comment]
I am thoroughly uninterested in maintaining these packages in nixpkgs if that means making them worse. If someone wants to comply with the review comments from this PR please feel free to pick up this code and follow up in a new PR.

**Comment by sliedes** [comment]
Has anything changed about how packages are done in pythonPackages since this? (I'm interested in having, and possibly packaging, austin, but this discussion gives me the feeling I'm not sure I'm up to it, as my first Python package...)

**Pull request closed without merging**
